mountpoint-s3-client/tests/put_object.rs

```diff
@@ -202,6 +206,46 @@ async fn test_put_object_abort(size: usize) {
     assert_eq!(uploads_in_progress, 0);
 }
 
+#[tokio::test]
+async fn test_put_object_write_cancelled() {
+    let (bucket, prefix) = get_test_bucket_and_prefix("test_put_object_write_cancelled");
+    let client = get_test_client();
+    let key = format!("{prefix}hello");
+
+    let mut request = client
+        .put_object(&bucket, &key, &Default::default())
+        .await
+        .expect("put_object should succeed");
+
+    // Complete one write to ensure the MPU was created
+    request.write(&[1, 2, 3, 4]).await.expect("write should succeed");
+
+    {
+        // Write at least `part_size` to ensure the copy is deferred.
+        let buffer = vec![0u8; client.part_size().unwrap()];
+        let mut write = request.write(&buffer);
```

Comment:
This is the comment in `aws_s3_meta_request_write`:
```
/* This write makes us ready to send (EOF, or we have enough data now to send at least 1 part) */
```
If we provide at least part size (or EOF), the CRT can just schedule an upload (which will read from the slice we passed in). If not, it will immediately copy the data in an internal buffer and return.